### ISSUE TITLE
Better memory management in V2Ray

### ISSUE THREAD

**Issue by v2ray** [v2ray/v2ray-core#67]
Golang is poor in memory allocation. Its performance is even worse than Java. (Benchmark: https://benchmarksgame.alioth.debian.org/u64q/performance.php?test=binarytrees)

V2Ray uses a customized memory management which pre-allocates memory for later use. It now uses a fixed strategy for all platforms. There is some room for platform dependent improvements.

**Comment by v2ray** [comment]
Fixed in 1.4. This is probably caused by GC in golang. Golang doesn't efficiently recycle all unused memory in time.

**Comment by pminmax945** [comment]
profile shows that  TimedUserSet.generateNewHashes consume most memory (even than alloc.newBufferPool),  I've found this is triggerd by  updateUserHash=>cacheDurationSec inner loop, reduce this from 120 to 50 can save about 10~15M (51M=>37M in my setup). could you please provide some design points for this cacheDurationSec variable?

**Comment by v2ray** [comment]
TL;DR: blame Golang for the memory consumption on generateNewHashes. Map in Golang consumes 10 times space more than an equivalent array.

The authentication of VMess (V2Ray's protocol) is based on the UTC time. V2Ray client generates authentication bits based on current UTC time with some other factors. For a better authentication performance, the server caches all possible authentication bits for a short period (defined by cacheDurationSec), using a hashmap (map in Golang).

cacheDurationSec means the maximum time different in seconds that is allowed between V2Ray client machine and server. cacheDurationSec = 120  allows maximum 2 minutes difference. If the time difference is over 2 minutes, all request will fail.

In your case, I assume you are setting a VMess protocol on server with "alterId" > 1000, which causes an over 50M memory allocation. However, the actual memory required should be far less. A map entry for storing the cache is as below (in userset.go):

```go
type Timestamp int64

type indexTimePair struct {
	index   int
	timeSec Timestamp
}

userHash   map[[16]byte]indexTimePair
```

One entry should only cost 4 + 8 + 16 = 28 bytes. The whole map should cost 28 * 1000 * 120 * 2 ~= 6.4M. However, Golang uses 50M as you measure.

To effectively reduce memory consumption in your case, I suggest to reduce the value of "alterId" to lower than 100, if that's what you concerns.

**Comment by pminmax945** [comment]
yes,this value is 1024 in config example. so this leads my setup 50M

**Comment by v2ray** [comment]
Completely fixed.
